enhancements/baremetal/debuggability-of-baremetal-ipi-deployment.md

```diff
@@ -0,0 +1,246 @@
+---
+title: debuggability-of-baremetal-ipi-deployment
+authors:
+  - "@stbenjam"
+reviewers:
+  - "@abhinavdahiya"
+  - "@dtantsur"
+  - "@enxebre"
+  - "@hardys"
+  - "@juliakreger"
+  - "@markmc"
+  - "@sadasu"
+approvers:
+  - TBD
+creation-date: 2020-05-15
+last-updated: 2020-05-15
+status: provisional
+see-also:
+- https://github.com/openshift/installer/pull/3535
+- https://github.com/openshift/enhancements/pull/212
+replaces:
+superseded-by:
+---
+
+# Improve debuggability of baremetal IPI deployment failures
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+1. Should the installer error or warn when compute replicas quantity is
+   not met, even if at lest 2 workers deployed (enough to get a
+   functional cluster)?
+2. In order to bubble up information about worker failures to the
+   installer, the most likely solution to me seems to use operator
+   status as Degraded, with a relevant error message. Could we mark
+   machine-api-operator when workers fail to roll out?
+
+## Summary
+
+In OpenShift 4.5, we improved the existing installer validations for
+baremetal IPI to identify early problems. Those include identifying
+duplicate baremetal host records, insufficient hardware resources to
+deploy the requested cluster size, reachability of RHCOS images, and
+networking misconfiguration such as overlapping networks or DNS
+misconfiguration.
+
+However, a variety of situations exist where deployments fail for
+reasons that were not preventable during the pre-install validations.
+These failures in baremetal IPI are hard to diagnose. Errors from
+baremetal-operator and ironic are often not presented to the user, and
+even when they are the installer doesn't provide context about what
+action to take.
+
+This enhancement request is a broad attempt at categorizing the types of
+deployment failures, and what information we could present to the user
+to make identifying root causes easier.
+
+## Motivation
+
+The goal of this enhancement is to improve the day 1 install experience
+and reduce the perception of complexity in baremetal IPI deployments.
+
+### Goals
+
+- Any deployment that ends in an unsuccessful install must provide the
+  user clear and actionable information to diagnose the problem.
+
+### Non-Goals
+
+- Addressing the underlying causes of the failures are not the goal of
+  this enhancement.
+
+## Proposal
+
+Broadly, deployments fail due to problems encountered during these
+installation activities:
+
+  - Pre-bootstrap (image downloading, manifest creation, etc)
+  - Infrastructure automation (Terraform)
+  - Bootstrap
+  - Bare Metal Host Provisioning (Control Plane and Workers)
```

Comment:
isn't the control plane piece of bare metal host provisioning captured by 2 steps earlier (terraform)?  You could update that line to clarify that provisioning of those first hosts is done at the terraform stage?

Reply:
It's explained a bit later, should I structure this differently?

I want to highlight terraform as a point of failure that we need to make more debuggable. However, provisioning issues are grouped together for workers and control plane since the root causes can be the same. The only difference is the communication path to the user is different (terraform vs MAO).

Reply:
OK, that's fair.  don't worry about changing this